### ISSUE TITLE
Excessive CPU usage

### ISSUE THREAD

**Issue by daniellm** [dotnet/orleans#2374]
Hi,

I'm load-testing a silo and am seeing excessive CPU usage. I'm calling a grain method in a loop. The method basically does nothing, and the amount of parameters and response being serialized back and forth is minimal. I'd expect seeing up to one CPU core fully loaded since only one client and grain instances are involved, but I'm actually seeing around 85% global CPU usage on my quad-core machine (8 logical cores).

What's even more puzzling is that running several loops in parallel, each calling a separate grain, does not generate much higher CPU usage and scales well. It seems as if there's some background work going on somewhere in the framework that gobbles up nearly all available CPU resources (when one grain is under load), but that yields these resources to grains if needed.

I reproduced this bug by modifying the "Tutorial.Minimal" Orleans code sample. Let me know if you want a branch to look at. Here's the code:

```
public interface IHello : Orleans.IGrainWithIntegerKey
{
    Task<string> SayHello(string msg);
}

class HelloGrain : Orleans.Grain, IHello
{
    public Task<string> SayHello(string msg)
    {
        return Task.FromResult(string.Format("You said {0}, I say: Hello!", msg));
    }
}

public interface IBenchmark : Orleans.IGrainWithIntegerKey
{
    Task Run(int instance, int count);
}

class BenchmarkGrain : Orleans.Grain, IBenchmark
{
    static long totalCount = 0;

    public async Task Run(int id, int targetRate)
    {
        DateTime start = DateTime.Now;
        var grain = GrainFactory.GetGrain<IHello>(id);

        int sent = 0;
        while ((DateTime.Now - start).TotalSeconds < 10)
            if (sent < (DateTime.Now - start).TotalSeconds * targetRate)
            {
                await grain.SayHello(sent++.ToString());
                Interlocked.Increment(ref totalCount);
            }
            else await Task.Delay(1);

        Console.WriteLine($"{DateTime.Now.ToString("HH:mm:ss.fff")} Grain #{id}, {sent} requests sent @ {Math.Round(sent / (DateTime.Now - start).TotalSeconds, 0)}/sec (target = {targetRate}/sec) total={totalCount}");
    }
}

static void DoSomeClientWork()
{
    // Orleans comes with a rich XML and programmatic configuration. Here we're just going to set up with basic programmatic config
    var config = Orleans.Runtime.Configuration.ClientConfiguration.LocalhostSilo(30000);
    GrainClient.Initialize(config);

    int concurrency = 1;
    int targetRate = 20000;

    while (concurrency-- > 0)
    {
        int grainId = concurrency; // capture variable
        Task.Run(() =>
        {
            var bench = GrainClient.GrainFactory.GetGrain<IBenchmark>(grainId);
            while (true)
                bench.Run(grainId, targetRate).Wait();
        });
    }
}
```

DoSomeClientWork() is called from the main thread. It calls the BenchmarkGrain which, in turn, calls the HelloGrain for 10 seconds in a loop. Raising the concurrency value results in better overall rate of grain calls with the same, high, CPU usage. Lowering the targetRate value will gradually lower CPU usage, but in a non-linear fashion.

Am I doing anything wrong that triggers this issue?

**Comment by gabikliot** [comment]
Your observations are mostly correct.
Orleans's CPU utilization is non linear with the load and the achieved throughput. We use some spin locks inside, and thus even at low/medium load CPU is relatively high.
I am a bit surprised at 85%, I would expect less, but maybe it is due to the 4 and not 8 real cores.

Indeed, as you push more and more load, the CPU will stay almost the same but the system will just be able to absorb more traffic.  We looked at way to change that, but **so far** any other implementation we tried, that utilizes less CPU at lower load, also had smaller throughout at higher loads. Therefore, for  maximizing throughput we found that to be the best strategy, **at least so far**.
Thus, we consider this not a bug, but a feature. 

A couple of comments on your benchmark:
1) I recommend not to content too much on the static counter with Interlocked. Count locally and add up to the global counter at the end of the benchmark.
2) Run on bigger machine.
3) Try scaling with multiple servers. That is where the real beauty comes into play. A lot of systems can achieve very good single box perf., but scaling linearly with added boxes is harder. After going from 1 to 2 you may even get less perf., (since half of the calls that were before local will now be remote), but from 2 to 3 to 4 to ... 100 you will get linear scaling (if you don't have bottleneck grains).
4) In your case the client is not dong much, so its less important but if the client is making a lot of calls, run it on a different box.

**Comment by sergeybykov** [comment]
To add to what @gabikliot already listed.
>I'd expect seeing up to one CPU core fully loaded since only one client and grain instances are involved

Messaging and serialization use multiple threads both on the client and the silo side. So don't expect a single core to be busy and others idle.

I wonder also how expensive `else await Task.Delay(1);` is when called in a tight loop.

**Comment by daniellm** [comment]
Gabi, Sergey, thank you for your as-usual excellent and prompt replies. Addressing your points:

> I recommend not to content too much on the static counter with Interlocked. Count locally and add up to the global counter at the end of the benchmark.

Yep, though Interlocked accounts for ~0.1% of the CPU, so no biggie.

> 2) Run on bigger machine.

Will do.

> 4) In your case the client is not dong much, so its less important but if the client is making a lot of calls, run it on a different box.

In my real-world scenario, the client is separate and sends requests over HTTP; we use something akin to [OrleansHttp](https://github.com/OrleansContrib/OrleansHttp) to listen to HTTP requests and invoke grain methods.

> Messaging and serialization use multiple threads both on the client and the silo side. So don't expect a single core to be busy and others idle.

Yes, but I do expect to see up to the equivalent of 1 cpu core capacity in use, even if spread across several cores (i.e. 25% global CPU usage for a quad-core machine).

> I wonder also how expensive else await Task.Delay(1); is when called in a tight loop.

It only occurs when the target rate is reached. In my example above the target rate of 20k requests/sec is not achievable on my hardware so it never sleeps.

Back to the problem at hand, we probably use Orleans in a different way than most, which makes this CPU usage strategy problematic for us. Here's our setup:

- We are self-hosted
- Each micro-service runs in its own Orleans cluster
- Each machine runs several silos, each from a different cluster (i.e. service)
- Most services have low traffic and are spread over 2-3 silos with low CPU usage

I worry that a single hotspot grain in some service will eat up all of the machine's CPU and affect other services running on that machine. We'll take precautions by limiting the number of CPU cores each silo/service accesses using CPU affinity and/or the MaxActiveThreads configuration, but it's less than ideal.

Do you think it's feasible to add another scheduling/throughput strategy that favors CPU usage over throughput, and choose the strategy through configuration? Would you accept a merge request for that? Can this be implemented in one/few places in the Orleans framework, or would we need to touch lots of code? Also, any thoughts about our deployment setup?

Thanks again!

**Comment by gabikliot** [comment]
I think there are 2 issues here, which are somewhat related at the end, but lets target them one by one, at least for now.
**The longer term proper way to do multi-tenancy**. We looked into that in the past a bit, but did not get very far way in my time, but maybe @sergeybykov and the team made/planning to make more progress on that. At any event, if you are really very interested, I would dedicate a whole separate issue just on that and do more brainstorming discussions. I am sure @sergeybykov will have ideas to share.

**For the more limited, tactical issue of non-linear CPU utilization**: I do think we can do something about it. As first step, we can do a couple of targeted small improvement that can help your scenario. We know of (maybe not all, but most) the places where spinning happens. Theses are usually already well isolated data structures that already sit behind interfaces. We can add a config variable "bool OptimizeForThroughput" that will be true by default and preserve the current behavior, but if set to false we will use alternative data structures (returned from a factory)  that utilize more locks. That way you will be able to configure your silos differently. I don't think it will be considered a "code duplication", but rather "optimizing for different setups". There will be a couple of places like this, but not too much. Some of them are direct Orleans data structures, but some are  .NET data structures. For example, we know that Blocking Collection uses spinning internally. So we will need to switch away those places too. But that could be a gradual iterative process, which you could drive. I will help identify those places and suggest alternatives, and you could try the alternatives data structures and see if that helps.

Once that is done, the 2nd phase **may** or **may not** be to look back on the code differences, perf. differences and whether we can do something more elegant to combine the both worlds. But as I said before, so far this was not very successful.

**Comment by sergeybykov** [comment]
I'm thinking maybe it'll be more efficient to discuss this over Skype.

I'm still a little puzzled by the 85% figure. We know people built and ran similar micro-servicy deployments, for example, with https://github.com/Microsoft/Yams, and CPU was never a problem.

>I'm actually seeing around 85% global CPU usage on my quad-core machine (8 logical cores).

Is this with hyper-threading on? Is it a desktop or laptop?

As @gabikliot mentioned, we could make spinning configurable. I'm curious to confirm first that this will actually help.

**Comment by dVakulen** [comment]
I think that that replacing of spinning with locks will just mitigate the problem for particular workload on cost of heavily reduced maximum throughput. Better solution would be either to use work-stealing queues with shared thread pool (obviously that would be harder to implement), or to start batching items processing on reaching CPU consumption threshold. 


>I worry that a single hotspot grain in some service will eat up all of the machine's CPU and affect other services running on that machine 

that shouldn't be a problem because running multiple orleans hosts pretty similar to running single host - they will consume 85% of CPU at median load, but should be able to handle significantly higher load due to great spinning reduction (while one service processes batch of items - another one will accumulate that batch).

**Comment by rrector** [comment]
While probably not a primary reason to do it, another benefit of being able to configure the spin-lock behavior is that sampling CPU profilers could actually be useful. Currently, trying to profile an Orleans app, so many of the samples are capturing the spin lock cycles, that the noise makes it very difficult to find actual hot paths in the results.

**Comment by ReubenBond** [comment]
Closing since Orleans no longer maintains threads for the Thread Pool or networking and instead leverages the .NET shared ThreadPool with entirely async IO.
